### PR TITLE
Chained Hash Map redesign

### PR DESCRIPTION
**Pull request by ThanGeo**
## Purpose of the Change and Brief Change Log
This pull requests refactors the ChainedHashMap so that it uses TupleBuffer in a more structured way. 

## Requirements
This branch builds on top of the paged_vector_redesign branch. When that branch is merged, then this one should follow. 

## Documentation
The ChainedHashMap did not use our self-contained, single-TupleBuffer-based data structure model. This redesign aims to fix this, enabling easier seralization, migration, representation and memory optimization for the ChainedHashMap. 

The HashMapSlice was also slightly redesigned, since it contains hash maps based on the number of workers and input streams. 

image

Previous PR: [https://github.com/nebulastream/nebulastream/pull/1667](url)

### CODE REVIEW

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:1]: we can probably pull this code into a shared test utility for other upcoming datastructures.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 1]
```diff
@@ -12,180 +12,1001 @@
...
+                    [&](const nautilus::val<AbstractHashMapEntry*>& newEntry)
+                    {
+                        const auto chainedEntry = static_cast<nautilus::val<ChainedHashMapEntry*>>(newEntry);
+                        const ChainedHashMapRef::ChainedEntryRef newEntryRef(chainedEntry, chainedHashMapBuffer, fieldKeys, fieldValues);
+                        newEntryRef.copyValuesToEntry(record, bm);
+                    },
+                    bm);
+            })));
+
+        readAtFn.emplace(engine->registerFunction(std::function(
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: what is read at? why is the index numerical? The testable interface should look like this:
```
testableMap.at(AnyVec key) -> std::optional<AnyVec> value
testableMap.put(AnyVec key, AnyVec value)
```

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:1]: In general the file is to much copy paste of the pagedvector.
Effectivly we want to compare map<AnyVec, AnyVec> with our chained hashmap and map<AnyVec,vector<AnyVec>> with our chm.

So you need a TestableHashMap that has at(key), put(key, value), and maybe a getAll to compare iterators

**Review by ls-1801 — CHANGES_REQUESTED** [review]
mostly looked at the tests so far

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:1]: Done, but i'm thinking maybe that eventually for more property testing on data structures we should find some better way to abstract/interface this anyvec code for correctness more easily.

Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:1]: done

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 1 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: you're right. Fixed it as per your suggestion

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp [hunk 2]
```diff
@@ -64,53 +66,114 @@ class ChainedHashMapEntry final : public AbstractHashMapEntry
...
+
+    [[nodiscard]] const Header& header() const { return *buffer.getAvailableMemoryArea<const Header>().data(); }
+
+    [[nodiscard]] std::span<ChainedHashMapEntry*> chains()
+    {
+        auto* data = buffer.getAvailableMemoryArea<uint8_t>().data();
+
+        /// NOLINTBEGIN(cppcoreguidelines-pro-type-reinterpret-cast, cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        auto* entries = reinterpret_cast<ChainedHashMapEntry**>(data + sizeof(Header));
+        return {entries, getNumberOfChains() + 1};
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:178]: All current callers allocate this buffer through `calculateBufferSizeFromBuckets/Chains`, so the layout should fit. Since `chains()` relies on that implicit pairing, could we add a `PRECONDITION` or `INVARIANT` in `init` checking that the buffer contains `Header` plus `numChains + 1` pointers? That would encode the layout contract and protect future serialization/migration call sites.

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp [hunk 3]
```diff
@@ -56,16 +61,10 @@ uint64_t calcCapacity(const uint64_t numberOfKeys, const double loadFactor)
...
-    , pageSize(pageSize)
-    , entrySize(entrySize)
-    , entriesPerPage(pageSize / entrySize)
-    , numberOfChains(calcCapacity(numberOfBuckets, assumedLoadFactor))
-    , entries(nullptr)
-    , mask(numberOfChains - 1)
-    , destructorCallBack(nullptr)
+void ChainedHashMap::init(TupleBuffer& tupleBuffer, const uint64_t entrySize, const uint64_t numberOfBuckets, const uint64_t pageSize)
 {
+    const uint64_t entriesPerPage = pageSize / entrySize;
```
Comment (ls-1801) [inline nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp:67]: `entrySize` is used as a divisor before the `entrySize > 0` precondition runs, so `init(..., entrySize = 0, ...)` hits division by zero instead of the intended contract failure. Could we move the precondition ahead of the `entriesPerPage` calculation?

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMapRef.cpp [hunk 4]
```diff
@@ -269,57 +268,75 @@ ChainedHashMapRef::EntryIterator ChainedHashMapRef::begin() const
     const nautilus::val<uint64_t> indexOnPage = 0;
     const nautilus::val<uint64_t> pageIndex = 0;
     const auto currentEntry = nautilus::invoke(
-        +[](const HashMap* hashMap, const uint64_t pageIndexVal, const uint64_t indexOnPageVal)
+        +[](TupleBuffer* tupleBuffer, const uint64_t pageIndexVal, const uint64_t indexOnPageVal)
         {
-            const auto& page = dynamic_cast<const ChainedHashMap*>(hashMap)->getPage(pageIndexVal);
+            const auto chm = ChainedHashMap::load(*tupleBuffer);
+            const auto& page = chm.getPage(pageIndexVal);
```
Comment (ls-1801) [inline nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMapRef.cpp:283]: An initialized empty map has no storage page yet, but `begin()` unconditionally loads page 0. The following regression test builds on this branch and currently aborts with exit code 134 (`terminate called without an active exception`):

```cpp
TEST(ChainedHashMapIteratorTest, emptyMapIsAnEmptyRange)
{
    constexpr uint64_t bufferSize = 4096;
    constexpr uint64_t numberOfBuckets = 1;
    constexpr uint64_t entrySize = sizeof(ChainedHashMapEntry);
    constexpr uint64_t entriesPerPage = 1;

    auto bufferManager = BufferManager::create(bufferSize, 16);
    auto hashMapBuffer = bufferManager->getUnpooledBuffer(
        ChainedHashMap::calculateBufferSizeFromBuckets(numberOfBuckets)).value();
    ChainedHashMap::init(hashMapBuffer, entrySize, numberOfBuckets, entrySize * entriesPerPage);

    auto engine = makeEngine(EngineMode::Interpreter);
    auto iterate = engine.registerFunction(std::function(
        [](nautilus::val<TupleBuffer*> buffer)
        {
            const ChainedHashMapRef ref(
                buffer, {}, {}, nautilus::val<uint64_t>{entriesPerPage}, nautilus::val<uint64_t>{entrySize});
            for (const auto entry : ref)
            {
                std::ignore = entry;
            }
        }));

    EXPECT_NO_THROW(iterate(&hashMapBuffer));
}
```

Could `begin()` return an iterator equal to `end()` when the map has zero records, without loading a page? This would also let the test remove its existing caller-side empty-map guard.

**Review by ls-1801 — CHANGES_REQUESTED** [review]

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp (hunk 2 above)
Comment (ThanGeo) [reply nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:178]: added a guard in init() that takes the chains into account for the size check

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp (hunk 3 above)
Comment (ThanGeo) [reply nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp:67]: moved the precondition before using the entrySize value

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMapRef.cpp (hunk 4 above)
Comment (ThanGeo) [reply nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMapRef.cpp:283]: merged all invoked by introducing a new struct that wraps the two unsigned long values together. Added a check in there about ```numPages = 0``` and returns ```end()``` if true.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 5]
```diff
@@ -12,180 +12,558 @@
...
+                }
+            }
+            return true;
+        });
+    return it == reference.end() ? nullptr : &it->second;
 }
 
-TEST_P(ChainedHashMapTest, update)
+/// Builds a reference association list tracking only the first-seen value per unique key,
+/// mirroring CHM's first-write-wins deduplication, and put()s every generated pair into the map under test.
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: Could you please document all behaviors in which our chained hash map differs from std::unordered_map?

This first-write-wins is new to me.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp
Comment (ls-1801) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp:1]: You mean code like this? Yeah, we could move them into the util

```c++
                Record record;
                /// static_val ensures each field iteration gets a distinct trace tag.
                for (nautilus::static_val<uint64_t> i = 0; i < dataTypes.size(); ++i)
                {
                    record.write(projections[i], buildVarVal(rec, i, dataTypes[i]));
                }
```

or the reverse:

```c++
for (nautilus::static_val<uint64_t> i = 0; i < fieldValues.size(); ++i)
                    {
                        const auto& fld = fieldValues[i];
                        storeVarValToAnyVec(out, static_cast<uint64_t>(i), valueRecord.read(fld.fieldIdentifier), fld.type);
                    }
```

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 6]
```diff
@@ -12,180 +12,558 @@
...
+        const auto valueProjections = std::vector<Record::RecordFieldIdentifier>(splitPoint, projections.end());
+        auto [fk, fv]
+            = ChainedEntryMemoryProvider::createFieldOffsets(createSchemaFromDataTypes(fieldTypes), keyProjections, valueProjections);
+        return {.keys = std::move(fk), .values = std::move(fv)};
     }
+};
 
-    /// Now we are searching for the entries and checking if the values are correct.
-    checkIfValuesAreCorrectViaFindEntry(hashMap, exactMap);
+using KeyValueReference = std::vector<std::pair<AnyVec, AnyVec>>;
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: What is the reason we are comparing against a std::vector?

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 7]
```diff
@@ -12,180 +12,558 @@
...
+        numEntriesPerPage,
+        fmt::join(fieldTypes, ", "));
+
+    auto bufferManager = BufferManager::create(bufferSize, pooledBufferCountFor(bufferSize));
+    TestableChainedHashMap chainedHashMap(fieldTypes, *bufferManager, mode, numberOfBuckets, numKeyFields, numEntriesPerPage);
+    const auto reference = buildReferenceAndPopulate(chainedHashMap, fieldTypes, numberOfItems);
+
+    NES_INFO("putAndLookupKeys: CHM has {} entries, {} unique keys", chainedHashMap.size(), reference.size());
+    RC_ASSERT(chainedHashMap.size() == reference.size());
+    verifyLookups(chainedHashMap, reference, fieldTypes);
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: You could run this in a loop.

HashMap m
for(num_iterations) {
m.put_some_value();
m.verify_lookups();
}

That way we also test interleaving, and it's not just write-then-read.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 8]
```diff
@@ -12,180 +12,558 @@
...
-        return ss.str();
-    });
+    }
+}
+
+/// Verify put()/at() round-trip: every reference key is found with its first-seen value, and independently
+/// generated keys that don't collide with the reference correctly report absent.
+void putAndLookupKeysProperty(EngineMode mode)
+{
+    const auto fieldTypes = *genDataTypeSchema(ALL_VALUE_TYPES, 1, MAX_SCHEMA_FIELDS);
```
Comment (ls-1801) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: Not quite sure what currently happens if we don't have any fields. I.e, just using the HashMap as a HashSet

File: nes-physical-operators/src/HashMapSlice.cpp [hunk 9]
```diff
@@ -21,61 +21,74 @@
...
+                createNewHashMapSliceArgs.valueSize,
+                createNewHashMapSliceArgs.numberOfBuckets,
+                createNewHashMapSliceArgs.pageSize);
+            hashMapBuffers.emplace_back(childBuffer.value());
+        }
+        else
+        {
+            throw BufferAllocationFailure("No unpooled TupleBuffer available for chained hash map child buffer!");
+        }
     }
```
Comment (ls-1801) [inline nes-physical-operators/src/HashMapSlice.cpp]: We decided to keep the vector in the slice for now! Is there a reason we do the allocation upfront now?
For a large number of worker threads and tiny windows, this can be quite slow and wasteful. I think we can get away with reserving the number of slots in the vector, but have each thread construct its buffer on demand. Maybe we switch to a std::vector<std::optional<TupleBuffer>> instead, or we keep the buffer default-constructed (empty).

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp [hunk 10]
```diff
@@ -64,53 +66,114 @@ class ChainedHashMapEntry final : public AbstractHashMapEntry
...
+            , storageSpaceIndex(TupleBuffer::INVALID_CHILD_BUFFER_INDEX_VALUE)
+            , varSizedSpaceIndex(TupleBuffer::INVALID_CHILD_BUFFER_INDEX_VALUE)
+        {
+        }
+    };
+
+    static_assert(std::is_trivially_destructible_v<Header>, "Header must be trivially destructible");
+
+    /// Helper util methods for safe access
+    [[nodiscard]] Header& header() { return *buffer.getAvailableMemoryArea<Header>().data(); }
```
Comment (ls-1801) [inline nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:168]: You should probably assert the `VALID_CHM` somewhere; might as well put it here.

**Review by ls-1801 — CHANGES_REQUESTED** [review]

**Comment by zeuchste** [comment]
We benchmark hash-join storage/processing variants on this codebase (edge-join design-space study) and ran this PR's head (d85467f6b5) through our workload battery — 23 points (window sizes 10ms–100s, duplicates 0.1–32/key, 2–32 cores, 0–8 co-running queries), keyed default join config, 2 reps each vs. base 609d016fcb, same nodes, checksum-verified. Sharing the results since they may be useful input:

**Correctness:** every run reproduced the exact golden checksum. 👍

**Throughput:** consistent **−10–16% steady-state time in the mid-range** (1s–100s windows, low-to-mid duplication) — the redesign measurably reduces the per-key allocation cost that dominates the current map on many-key workloads. Three regressions at the extremes:
- 10ms windows (2.5M windows): **+26%** — per-window map setup appears heavier;
- 2-core pinned: +9%, unpinned 64-CPU: +7%;
- benefit vanishes at heavy duplication (32/key: −2%).

**Tail latency under concurrency:** with ≥4 co-running queries, trigger→sink latency explodes: p50 34ms → 7.3s (n=4) / 15.2s (n=8), p99 up to 66s, while throughput still improves. Caveat: we could not pin memory config equal (see below), so the new default memory bound may be a confound — but worth a look before merge.

Two harness observations from testing (possibly main, not this PR):
1. Passing `--worker.total_memory_in_bytes=<any value>` through systest's `--` passthrough exits with code 232 and no error message.
2. When the disable-config filters out all selected tests, systest prints its banner and exits with success-like output — an A/B run of ours silently executed zero queries.

Happy to share the full per-point table and repro scripts.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 5 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: Added some elaboration on how the HashMap works overall

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 6 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: Switched to comparing against a std::unordered_map

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 7 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: Done

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp (hunk 8 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: HashSet was already supported internally, i added a property test for it too

File: nes-physical-operators/src/HashMapSlice.cpp (hunk 9 above)
Comment (ThanGeo) [reply nes-physical-operators/src/HashMapSlice.cpp]: switched back to lazy initialization of buffers. I associated a new vector of flags along with the vector of tuplebuffers to keep track of what buffers have been initialized. I know it's not the prettiest solution, but If i had to switch to storing children buffers instead in order to use the INVALID_CHILD_BUFFER_INDEX placeholder here it would re-introduce the complexity that we got rid of by reverting back to the vector again.

File: nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp (hunk 10 above)
Comment (ThanGeo) [reply nes-nautilus/include/Interface/HashMap/ChainedHashMap/ChainedHashMap.hpp:168]: I assert it in the load() method of the chained hash map. The status flag is always the first field in the header(), so if you load a trash/unrelated buffer then you can catch this in the load() method. The header() is only used internally, after a load

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp [hunk 11]
```diff
@@ -80,18 +79,34 @@ ChainedHashMap::ChainedHashMap(uint64_t entrySize, const uint64_t numberOfBucket
...
+    PRECONDITION(
+        tupleBuffer.getBufferSize() >= sizeof(Header) + (numberOfChains + 1) * sizeof(ChainedHashMapEntry*),
+        "Buffer of size {} is not big enough to hold the header ({} bytes) plus {} chain pointers ({} bytes each)",
+        tupleBuffer.getBufferSize(),
+        sizeof(Header),
+        numberOfChains + 1,
+        sizeof(ChainedHashMapEntry*));
+
+    /// Create object
+    ChainedHashMap chm(tupleBuffer);
```
Comment (keyseven123) [inline nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp]: Use `{}` for constructors

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp [hunk 12]
```diff
@@ -106,156 +121,277 @@ ChainedHashMap::ChainedHashMap(const uint64_t keySize, const uint64_t valueSize,
...
+        const uint64_t lastPageSize = lastPage.getBufferSize();
+        const uint64_t lastPageNumTuples = lastPage.getNumberOfTuples();
+        needNewPage = (lastPageNumTuples + neededSize > lastPageSize);
+    }
+
+    if (needNewPage)
+    {
+        /// Currently if a single varsized entry can not fit in one page we throw an exception
+        PRECONDITION(
+            neededSize <= bufferProvider->getBufferSize(),
```
Comment (keyseven123) [inline nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp]: But should we not rather use the unpooled buffers instead of having this hard limit of varsized data in hash maps?

File: nes-nautilus/include/Interface/HashMap/HashMapRef.hpp [hunk 13]
```diff
@@ -59,7 +61,7 @@ class HashMapRef
     virtual nautilus::val<AbstractHashMapEntry*> findEntry(const nautilus::val<AbstractHashMapEntry*>& otherEntry) = 0;
 
 protected:
-    nautilus::val<HashMap*> hashMapRef;
+    nautilus::val<TupleBuffer*> tupleBuffer;
```
Comment (keyseven123) [inline nes-nautilus/include/Interface/HashMap/HashMapRef.hpp]: please rename this member. We also discussed that we want to go away from `tupleBuffer`. What about simply `buffer` or `hashmap`?

File: nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+#include <static.hpp>
+
+namespace NES::TestUtils
+{
+
+std::shared_ptr<BufferManager> createBufferManager(const uint64_t bufferSize, const uint64_t numberOfBuffers)
+{
+    constexpr double unpooledMemoryFraction = 0.9;
+    constexpr BufferAlignment alignment{64};
+    const size_t totalMemoryInBytes = 10 * numberOfBuffers * bufferSize;
```
Comment (keyseven123) [inline nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp:56]: Is there a reason for this `10`? If so then add a comment. Otherwise, people are tempted to remove it

File: nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+Schema<QualifiedUnboundField, Ordered> createSchemaFromDataTypes(const std::vector<DataType>& dataTypes)
+{
+    return views::enumerate(dataTypes)
+        | std::views::transform(
+               [](const auto& pair)
+               {
+                   const auto& [typeIdx, dataType] = pair;
+                   const auto name = Record::RecordFieldIdentifier(Identifier::parse("field" + std::to_string(typeIdx)));
```
Comment (keyseven123) [inline nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp]:
```suggestion c++
const Record::RecordFieldIdentifier name{Identifier::parse("field" + std::to_string(typeIdx))};
```

File: nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+                   return QualifiedUnboundField{name, dataType};
+               })
+        | std::ranges::to<Schema<QualifiedUnboundField, Ordered>>();
+}
+
+rc::Gen<std::vector<DataType>> genDataTypeSchema(std::span<const DataType::Type> pool, size_t minFields, size_t maxFields)
```
Comment (keyseven123) [inline nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp]: not a fan of naming it pool. What about `possibleDataTypes`?

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 17]
```diff
@@ -12,180 +12,671 @@
...
+    types.reserve(fields.size());
+    for (const auto& field : fields)
+    {
+        names.push_back(field.fieldIdentifier);
+        types.push_back(field.type);
+    }
+    return {std::move(names), std::move(types)};
+}
+
+class TestableChainedHashMap
```
Comment (keyseven123) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: I think we should move this into its own class as it is quite large. We already have a custom `DataStructureTestUtils` so another `TestableChainedHashmap` is fine. Then this file here can take care of only different testing cases  / scenarios.

File: nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp [hunk 18]
```diff
@@ -12,180 +12,671 @@
...
+        if (not chainedHashMapBufferOpt.has_value())
+        {
+            throw BufferAllocationFailure(
+                "No unpooled TupleBuffer of size {} available!", ChainedHashMap::calculateBufferSizeFromBuckets(numberOfBuckets));
+        }
+        chainedHashMapBuffer = chainedHashMapBufferOpt.value();
+        ChainedHashMap::init(chainedHashMapBuffer, entrySize, numberOfBuckets, pageSize);
+
+        const auto numKeyProjections = keyDataTypes.size();
+        const auto keyProjections = std::vector<Record::RecordFieldIdentifier>(
```
Comment (keyseven123) [inline nes-nautilus/tests/UnitTests/ChainedHashMapTest.cpp]: So we assume that the projections for the key are always the first in the projections? Not sure if this is the best we can do... Can we at least have a precondition that checks if the projections is large enough?

File: nes-physical-operators/src/HashMapSlice.cpp (hunk 9 above)
Comment (ls-1801) [reply nes-physical-operators/src/HashMapSlice.cpp]: sounds like what a std::vector<std::optional<TupleBuffer>> would also do, without a second allocation

File: nes-nautilus/tests/UnitTests/PagedVectorTest.cpp [hunk 19]
```diff
@@ -59,27 +60,12 @@
 #include <fmt/ranges.h>
 #include <rapidcheck/gtest.h>
 
-/// NOLINTBEGIN(misc-include-cleaner, bugprone-unchecked-optional-access)
+/// NOLINTBEGIN(misc-include-cleaner, bugprone-unchecked-optional-access, google-build-using-namespace)
```
Comment (keyseven123) [inline nes-nautilus/tests/UnitTests/PagedVectorTest.cpp]: I am not a fan of this, as these are good clang-tidy warnings. Is there a bigger reason why you disabled them?

File: nes-physical-operators/include/Join/NestedLoopJoin/NLJSlice.hpp [hunk 20]
```diff
@@ -68,6 +68,7 @@ class NLJSlice final : public Slice
     void combinePagedVectors();
 
 private:
+    /// This does not really follow our compact-buffer data structure logic
```
Comment (keyseven123) [inline nes-physical-operators/include/Join/NestedLoopJoin/NLJSlice.hpp]: So should we not change it then?

**Review by ls-1801 — APPROVED** [review]

File: nes-physical-operators/src/Aggregation/AggregationOperatorHandler.cpp [hunk 21]
```diff
@@ -104,13 +105,18 @@ void AggregationOperatorHandler::triggerSlices(
         /// - all pointers to all hashmaps of the window to be triggered
         /// - a new hashmap for the probe operator, so that we are not overwriting the thread local hashmaps
         /// - size of EmittedAggregationWindow
```
Comment (keyseven123) [inline nes-physical-operators/src/Aggregation/AggregationOperatorHandler.cpp]: Please update this comment

File: nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp [hunk 22]
```diff
@@ -49,53 +49,6 @@
...
-    /// ReSharper disable once CppPassValueParameterByConstReference
-    /// NOLINTBEGIN(performance-unnecessary-value-param)
-    auto* const operatorHandler = dynamic_cast<HJOperatorHandler*>(
-        nautilus::details::RawValueResolver<OperatorHandler*>::getRawValue(executionCtx.getGlobalOperatorHandler(operatorHandlerId)));
-    if (operatorHandler->wasSetupCalled(joinBuildSide))
-    {
-        return;
-    }
-
-    const auto cleanupStateNautilusFunction
```
Comment (keyseven123) [inline nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp:69]: so nice to see this finally gone :)

File: nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp [hunk 23]
```diff
@@ -133,32 +90,40 @@ void HJBuildPhysicalOperator::execute(ExecutionContext& ctx, Record& record) con
...
+                    +[](TupleBuffer* hashMapBuf, uint32_t* valueMemArea, AbstractBufferProvider* bufferProvider, uint64_t tupleSize) -> void
                     {
                         if (auto pagedVectorBuffer = bufferProvider->getUnpooledBuffer(PagedVector::getMainBufferSize()))
                         {
-                            /// initialize paged vector buffer
                             PagedVector::init(pagedVectorBuffer.value(), bufferProvider->getBufferSize(), tupleSize);
-                            /// @warning: this will be refactored again during the ChainedHashMap refactor
-                            new (pagedVectorBufferMemArea) TupleBuffer(pagedVectorBuffer.value());
+                            auto childIndex = hashMapBuf->storeChildBuffer(pagedVectorBuffer.value());
+                            *valueMemArea = childIndex.getRawIndex();
```
Comment (keyseven123) [inline nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp:103]: Please add here why we are storing the raw index at it.

**Review by keyseven123 — CHANGES_REQUESTED** [review]
Generally, I am really happy that we finally get rid of the cleanup methods.

File: nes-physical-operators/src/HashMapSlice.cpp (hunk 9 above)
Comment (ThanGeo) [reply nes-physical-operators/src/HashMapSlice.cpp]: Should I switch to that?

File: nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp (hunk 14 above)
Comment (ThanGeo) [reply nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp:56]: it is the exact same thing being used in a lot of other tests as well, e.g. EmitPhysicalOperatorTest.cpp:72 (commited by you).

File: nes-physical-operators/include/Join/NestedLoopJoin/NLJSlice.hpp (hunk 20 above)
Comment (ThanGeo) [reply nes-physical-operators/include/Join/NestedLoopJoin/NLJSlice.hpp]: Not really because the NLJSlice is not a data structure. It added significant complexity and overhead trying to model it that way with no real benefits

File: nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp (hunk 23 above)
Comment (ThanGeo) [reply nes-physical-operators/src/Join/HashJoin/HJBuildPhysicalOperator.cpp:103]: This gets overwritten by the ChildBufferIndex PR. But the idea is you only care about the number of the child buffer and write it in a nautilus-wrappable type

File: nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp (hunk 12 above)
Comment (ThanGeo) [reply nes-nautilus/src/Interface/HashMap/ChainedHashMap/ChainedHashMap.cpp]: You're right, fixed it in the same way as in the PagedVector (#1814)

File: nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp (hunk 16 above)
Comment (ThanGeo) [reply nes-nautilus/tests/TestUtils/src/DataStructureTestUtils.cpp]: kept the pool but made it more explicit

File: nes-nautilus/tests/UnitTests/PagedVectorTest.cpp (hunk 19 above)
Comment (ThanGeo) [reply nes-nautilus/tests/UnitTests/PagedVectorTest.cpp]: fixed it

**Review by keyseven123 — APPROVED** [review]
